doc/weakly_supervised.rst

```diff
@@ -0,0 +1,316 @@
+   to specify pairs is more efficient
+
+
+2D array of indicators + preprocessor
+-------------------------------------
+
+Instead of forming each point in each tuple, a more efficient representation
+would be to keep the dataset of points ``X`` aside, and just represent tuples
+as a collection of tuples of *indices* from the points in ``X``. Since we loose
+the feature dimension there, the resulting array is 2D.
+
+.. rubric:: Example: An equivalent representation of the above pairs would be:
+
+>>> X = np.array([[-0.12, -1.21, -0.20],
+>>>               [+0.05, -0.19, -0.05],
+>>>               [-2.16, +0.11, -0.02],
+>>>               [+1.58, +0.16, +0.93],
+>>>               [+0.89, -0.34, +2.41]])
+>>>
+>>> tuples_indices = np.array([[0, 1],
+>>>                            [2, 3],
+>>>                            [3, 4],
+>>>                            [0, 2]])
+>>> y = np.array([-1, 1, 1, -1])
+
+In order to fit metric learning algorithms with this type of input, we need
+to give the original dataset of points ``X`` to the estimator so that it
+knows what point the indices refer to. We do this when initializing the
+estimator, through the argument `preprocessor`.
+
+.. rubric:: Example:
+
+>>> from metric_learn import MMC
+>>> mmc = MMC(preprocessor=X)
+>>> mmc.fit(pairs_indice, y)
+
+
+.. note::
+
+   Instead of an array-like, you can give a callable in the argument
+   ``preprocessor``, which will go fetch and form the tuples. This allows to
+   give more general indicators than just indices from an array (for instance
+   paths in the filesystem, name of records in a database etc...) See section
+   :ref:`preprocessor` for more details on how to use the preprocessor.
+
+
+Scikit-learn compatibility
+==========================
+
+Weakly supervised estimators are compatible with scikit-learn routines for
+model selection (grid-search, cross-validation etc). See the scoring section
+for more details on what scoring is used in the case of Weakly Supervised
+Metric Learning.
+
+.. rubric:: Example
+
+>>> from metric_learn import MMC
+>>> from sklearn.datasets import load_iris
+>>> from sklearn.model_selection import cross_val_score
+>>> rng = np.random.RandomState(42)
+>>> X, _ = load_iris(return_X_y=True)
+>>> # let's sample 30 random pairs and labels of pairs
+>>> pairs_indices = rng.randint(X.shape[0], size=(30, 2))
+>>> y = rng.randint(2, size=30)
+>>> mmc = MMC(preprocessor=X)
+>>> cross_val_score(mmc, pairs_indices, y)
+
+Scoring
+=======
+
+Some default scoring are implemented in metric-learn, depending on which
+kind of tuples you work on. See the docstring of the `score` method of the
+estimator you use.
+
+
+Algorithms
+==================
+
+Note that each weakly-supervised algorithm has a supervised version of the form
+`*_Supervised` where similarity tuples are generated from
+the labels information and passed to the underlying algorithm.
+
+.. todo:: add more details on `_Supervised` classes
+
+1. ITML
```

Comment:
the numbering of the section for each algorithm has an issue (e.g., shows 3.4.1. 1. ITML)

Reply:
Thanks, done